node/node.go

```diff
@@ -0,0 +1,445 @@
+	w              *worker.Worker
+
+	autopilotAddr     *string
+	autopilotPassword *string
+	ap                *autopilot.Autopilot
+
+	apiAddr     string
+	apiPassword string
+	auth        func(http.Handler) http.Handler
+	dir         string
+	l           net.Listener
+	walletKey   consensus.PrivateKey
+
+	mux TreeMux
+	srv http.Server
+
+	cleanupFuncs []func() error
+}
+
+// NewNode creates a new, empty node without any components. The components are
+// then added through the various exposed methods. For a more detailed
+// description take a look at the Node's type documentation.
+func NewNode(apiAddr, apiPassword, dir string, uiHandler http.Handler, wk consensus.PrivateKey) (*Node, error) {
+	// Start listening.
+	l, err := net.Listen("tcp", apiAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	mux := TreeMux{
+		H:   uiHandler,
+		Sub: make(map[string]http.Handler),
+	}
+
+	// Overwrite APIAddr now that we know the exact addr from the listener.
+	return &Node{
+		apiAddr:     "http://" + l.Addr().String(),
+		apiPassword: apiPassword,
+		auth:        jape.BasicAuth(apiPassword),
+		dir:         dir,
+		l:           l,
+		mux:         mux,
+		srv:         http.Server{Handler: mux},
+		walletKey:   wk,
+	}, nil
+}
+
+// APIAddress returns the node's API address.
+func (n *Node) APIAddress() string {
+	return n.apiAddr
+}
+
+// Close shuts down the API as well as the components that were created.
+// Remote components that were added instead of created will remain active.
+func (c *Node) Shutdown(ctx context.Context) error {
+	defer c.l.Close() // close last
+
+	if err := c.srv.Shutdown(ctx); err != nil {
+		return err
+	}
+	var errs error
+	for _, f := range c.cleanupFuncs {
+		err := f()
+		if err == nil {
+			continue
+		} else if errs == nil {
+			errs = err
+		} else {
+			errs = fmt.Errorf("%w; %s", errs, err.Error())
+		}
+	}
+	return errs
+}
+
+// Serve starts serving the API. This is blocking and can be interrupted by
+// calling 'Close'.
+func (n *Node) Serve() error {
+	return n.srv.Serve(n.l)
+}
+
+// AddBus adds an already running, remote bus to the node.
+func (n *Node) AddBus(busAddr, busPassword string) error {
+	if n.busAddr != nil || n.busPassword != nil {
+		return errors.New("cluster already contains a bus")
+	}
+	// Set bus address and password to remote bus.
+	n.busAddr = &busAddr
+	n.busPassword = &busPassword
+	return nil
+}
+
+// CreateBus creates a new bus.
+func (n *Node) CreateBus(bootstrap bool, gatewayAddr string) error {
+	if n.busAddr != nil || n.busPassword != nil {
+		return errors.New("cluster already contains a bus")
+	}
+	b, cleanup, err := newBus(BusConfig{
+		Bootstrap:   bootstrap,
+		GatewayAddr: gatewayAddr,
+	}, n.dir, n.walletKey)
+	if err != nil {
+		return err
+	}
+	// Set bus address and password to local bus.
+	n.busAddr = &n.apiAddr
+	n.busPassword = &n.apiPassword
+	n.cleanupFuncs = append(n.cleanupFuncs, cleanup)
+	n.bus = b
+	n.mux.Sub["/api/store"] = &TreeMux{H: n.auth(bus.NewServer(b))}
```

Comment:
I guess this endpoint should be `/api/bus` now

Reply:
ok